### ISSUE TITLE
Add option to disable "Master authorized networks" in `google_container_cluster`

### ISSUE THREAD

**Issue by paulpalamarchuk** [terraform-providers/terraform-provider-google#4093]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Terraform Version

Terraform v0.11.14
Terraform v0.12.4

### Affected Resource(s)

* provider.google-beta 2.11.0
* provider.google 2.11.0

### Terraform Configuration Files

Original config (TFv0.11):

```tf
locals {
  cluster_type = "simple-private-znn"
}

provider "google-beta" {
  version = "2.11.0"
  region  = "${var.region}"
}

provider "google" {
  version = "2.11.0"
  region  = "${var.region}"
}

provider "kubernetes" {
  version = "1.8.1"
}

data "google_compute_subnetwork" "subnetwork" {
  name    = "${var.subnetwork}"
  project = "${var.project_id}"
  region  = "${var.region}"
}

module "gke" {
  source                  = "../../modules/private-cluster/"
  project_id              = "${var.project_id}"
  name                    = "${local.cluster_type}-cluster${var.cluster_name_suffix}"
  regional                = false
  region                  = "${var.region}"
  zones                   = "${var.zones}"
  network                 = "${var.network}"
  subnetwork              = "${var.subnetwork}"
  ip_range_pods           = "${var.ip_range_pods}"
  ip_range_services       = "${var.ip_range_services}"
  service_account         = "${var.compute_engine_service_account}"
  enable_private_endpoint = true
  enable_private_nodes    = true
  master_ipv4_cidr_block  = "172.16.0.0/28"

  master_authorized_networks_config = [{}]
}

data "google_client_config" "default" {}
```

Trying to disable "Master authorized networks":

```tf
locals {
  cluster_type = "simple-private-znn"
}

provider "google-beta" {
  version = "2.11.0"
  region  = "${var.region}"
}

provider "google" {
  version = "2.11.0"
  region  = "${var.region}"
}

provider "kubernetes" {
  version = "1.8.1"
}

data "google_compute_subnetwork" "subnetwork" {
  name    = "${var.subnetwork}"
  project = "${var.project_id}"
  region  = "${var.region}"
}

module "gke" {
  source                  = "../../modules/private-cluster/"
  project_id              = "${var.project_id}"
  name                    = "${local.cluster_type}-cluster${var.cluster_name_suffix}"
  regional                = false
  region                  = "${var.region}"
  zones                   = "${var.zones}"
  network                 = "${var.network}"
  subnetwork              = "${var.subnetwork}"
  ip_range_pods           = "${var.ip_range_pods}"
  ip_range_services       = "${var.ip_range_services}"
  service_account         = "${var.compute_engine_service_account}"
  enable_private_endpoint = true
  enable_private_nodes    = true
  master_ipv4_cidr_block  = "172.16.0.0/28"

  master_authorized_networks_config = []
}

data "google_client_config" "default" {}
```

### Debug Output

$ terraform apply

```sh
Warning: module.gke.data.google_container_engine_versions.region: "region": [DEPRECATED] Use location instead



Warning: module.gke.data.google_container_engine_versions.zone: "zone": [DEPRECATED] Use location instead



Warning: module.gke.google_container_cluster.zonal_primary: "zone": [DEPRECATED] Use location instead



Warning: module.gke.google_container_node_pool.zonal_pools: "zone": [DEPRECATED] use location instead



random_string.cluster_service_account_suffix: Refreshing state... (ID: none)
data.google_client_config.default: Refreshing state...
data.google_compute_subnetwork.subnetwork: Refreshing state...
data.google_container_engine_versions.region: Refreshing state...
data.google_compute_subnetwork.gke_subnetwork: Refreshing state...
data.google_compute_zones.available: Refreshing state...
data.google_client_config.default: Refreshing state...
data.google_compute_network.gke_network: Refreshing state...
random_shuffle.available_zones: Refreshing state... (ID: -)
data.google_container_engine_versions.zone: Refreshing state...
google_container_cluster.zonal_primary: Refreshing state... (ID: simple-private-znn-cluster)
google_container_node_pool.zonal_pools: Refreshing state... (ID: us-central1-a/simple-private-znn-cluster/default-node-pool)
null_resource.wait_for_zonal_cluster: Refreshing state... (ID: 2913453794839032241)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ module.gke.google_container_cluster.zonal_primary
      master_authorized_networks_config.#: "1" => "0"


Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.gke.google_container_cluster.zonal_primary: Modifying... (ID: simple-private-znn-cluster)
  master_authorized_networks_config.#: "1" => "0"

Error: Error applying plan:

1 error occurred:
        * module.gke.google_container_cluster.zonal_primary: 1 error occurred:
        * google_container_cluster.zonal_primary: googleapi: Error 400: Must specify a field to update., badRequest





Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

### Expected Behavior

Disable "Master authorized networks" for the existing cluster

### Actual Behavior

Terraform failed to apply new configuration

### Steps to Reproduce

1. `terraform apply` - original config + .tfvars and module files
2. Change `master_authorized_networks_config` from `[{}]` to `[]`
3. `terraform apply` 

### Important Factoids

Disabled "Master authorized networks" with gcloud-cli:

`gcloud container clusters update simple-private-znn-cluster --zone=us-central1-a --no-enable-master-authorized-networks`

Used [Terraform GKE](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine) [v3.0.0](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/v3.0.0) and [v4.0.0](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/v4.0.0) as core module

### References

* [#154](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/154)

**Comment by rsawer** [comment]
I have the same issue with terraform-0.12.8 + terraform-provider-google_v2.14.0_x4

It's also simply to reproduce by removing master_authorized_networks_config from you k8s cluster declaration.

It produces an empty update request to google api:

[DEBUG] Google API Request Details:
 ---[ REQUEST ]---------------------------------------
PUT /v1beta1/projects/<project>/locations/<location>/clusters/<cluster name>?alt=json&prettyPrint=false HTTP/1.1
Host: container.googleapis.com
User-Agent: google-api-go-client/0.5 Terraform/0.12.6 (+https://www.terraform.io) terraform-provider-google/2.14.0
Content-Length: 14
Content-Type: application/json
Accept-Encoding: gzip

{
  "update": {}
}

-----------------------------------------------------
[DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 400 Bad Request
Alt-Svc: quic=":443"; ma=2592000; v="46,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Tue, 10 Sep 2019 05:55:23 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "error": {
    "code": 400,
    "message": "Must specify a field to update.",
    "errors": [
    {
      "message": "Must specify a field to update.",
      "domain": "global",
      "reason": "badRequest"
    }
    ],
    "status": "INVALID_ARGUMENT"
    }
 }

**Comment by prateek2408** [comment]
I am facing the same issue where I am unable to disable master authorized networks,

Any updates on this would be very helpful

**Comment by rsawer** [comment]
there's a simple workaround for public k8s clusters, just specify 0.0.0.0/0 as a authorized address, you will have your terraform passing without any problems and the result will be the same as not specifying master_authorized_networks_config block - the cluster will accept connections from any address

**Comment by prateek2408** [comment]
I tried but its gives me an error - 

Error: Unsupported argument

  on kubernetes.tf line 37, in resource "google_container_cluster" "testcluster":
  37:    master_authorized_networks_config  = [

An argument named "master_authorized_networks_config" is not expected here.
Did you mean to define a block of type "master_authorized_networks_config"?

**Comment by rsawer** [comment]
master_authorized_networks_config {
    cidr_blocks {
      cidr_block = "0.0.0.0/0"
      display_name = "Public access"
    }
  }

**Comment by prateek2408** [comment]
ah that is a dictionary and not a list, It works however now i get - 

Error: Error waiting for updating GKE cluster master authorized networks: Invalid master authorized networks: network "0.0.0.0/0" is not a reserved network, which is required for private endpoints.

**Comment by rsawer** [comment]
OK it seem's that you have a private cluster, specify your private cidr block there instead of 0.0.0.0/0

**Comment by prateek2408** [comment]
So it turns out that when you turn this feature on along with the cluster being private then you cannot put a public CIDR like 0.0.0.0 because it demands private endpoints,

So you have to disable the feature entirely which you can do via this command after the cluster gets created- 

gcloud container clusters update [CLUSTER_NAME] \
--no-enable-master-authorized-networks

More info about this here- https://cloud.google.com/kubernetes-engine/docs/how-to/authorized-networks

**Comment by rsawer** [comment]
as I said - the way I presented is only a workaround for configuring your GCP K8s cluster with terraform. For public cluster specifying 0.0.0.0/0 in cidr_block is equivalent for disabling master_authorized_networks, and specifying private cidr blocks (10..., 172..., etc) will be an equivalent for doing the same for private cluster

**Comment by megan07** [comment]
Thanks @paulpalamarchuk for opening this issue!  When I upgraded to the most recent version of the provider I was not able to reproduce the error, it looks like it was fixed with https://github.com/terraform-providers/terraform-provider-google/pull/4446 and was released in the 2.15.0 release.  I will close this, but please upgrade and open a new issue if you continue to see any errors.
